src/dbnode/storage/bootstrap/bootstrapper/fs/source.go

```diff
@@ -360,123 +480,148 @@ func (s *fileSystemSource) tagsFromTagsIter(
-				case series.CacheAll:
-					validateErr = r.Validate()
-				case series.CacheAllMetadata:
+			numEntries := r.Entries()
+			for i := 0; err == nil && i < numEntries; i++ {
+				switch run {
+				case bootstrapDataRunType:
+					err = s.readNextEntryAndRecordBlock(r, runResult, start, blockSize, shardResult,
+						shardRetriever, blockPool, seriesCachePolicy)
+				case bootstrapIndexRunType:
+					// We can just read the entry and index if performing an index run
+					err = s.readNextEntryAndIndex(r, runResult, indexBlockSegment)
+				default:
+					// Unreachable unless an internal method calls with a run type casted from int
+					panic(fmt.Errorf("invalid run type: %d", run))
+				}
+			}
+
+			if err == nil {
+				// Validate the read results
+				var validateErr error
+				switch run {
+				case bootstrapDataRunType:
+					switch seriesCachePolicy {
+					case series.CacheAll:
+						validateErr = r.Validate()
+					case series.CacheAllMetadata:
+						validateErr = r.ValidateMetadata()
+					default:
+						err = fmt.Errorf("invalid series cache policy: %s", seriesCachePolicy.String())
+					}
+				case bootstrapIndexRunType:
 					validateErr = r.ValidateMetadata()
 				default:
-					err = fmt.Errorf("invalid series cache policy: %s", seriesCachePolicy.String())
+					// Unreachable unless an internal method calls with a run type casted from int
+					panic(fmt.Errorf("invalid run type: %d", run))
+				}
+				if validateErr != nil {
+					err = fmt.Errorf("data validation failed: %v", validateErr)
 				}
-			case bootstrapIndexRunType:
-				validateErr = r.ValidateMetadata()
-			default:
-				// Unreachable unless an internal method calls with a run type casted from int
-				panic(fmt.Errorf("invalid run type: %d", run))
 			}
-			if validateErr != nil {
-				err = fmt.Errorf("data validation failed: %v", validateErr)
+
+			if err == nil && run == bootstrapIndexRunType {
+				// Mark index block as fulfilled
+				fulfilled := result.ShardTimeRanges{
+					shard: xtime.Ranges{}.AddRange(timeRange),
+				}
+				err = runResult.index.IndexResults().MarkFulfilled(start, fulfilled,
+					ns.Options().IndexOptions())
 			}
-		}
 
-		if err == nil && run == bootstrapIndexRunType {
-			// Mark as fulfilled
-			fulfilled := result.ShardTimeRanges{
-				shard: xtime.NewRanges(timeRange),
+			if err == nil {
+				remainingRanges.Subtract(result.ShardTimeRanges{
+					shard: xtime.Ranges{}.AddRange(timeRange),
+				})
+			} else {
+				s.log.Errorf("%v", err)
+				timesWithErrors = append(timesWithErrors, timeRange.Start)
 			}
-			runResult.Lock()
-			err = runResult.index.IndexResults().MarkFulfilled(start, fulfilled,
-				ns.Options().IndexOptions())
-			runResult.Unlock()
 		}
+	}
 
-		if err == nil {
-			tr = tr.RemoveRange(timeRange)
-		} else {
-			s.log.Errorf("%v", err)
-			timesWithErrors = append(timesWithErrors, timeRange.Start)
+	incremental := runOpts.Incremental()
+	anyRemaining := remainingRanges.IsEmpty()
```

Comment:
maybe `noneRemaining` would be a better name?

Reply:
Sure thing.